Improve and add exceptions

leandro-lucarella-frequenz opened the pull request:
- Properly await for service task
- Make output more verbose for darglint
- Add base exception
- Make `send()` raise a `SenderError`
- Make receivers raise `ReceiverError`
- Move exceptions into its own module
- Add a `ReceiverInvalidatedError`
- Don't raise exceptions in Receiver.ready()

Fixes #52.

llucax committed:
Add base exception

All exceptions from this library inherit from this exception.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Make receivers raise ReceiverError

The `Receiver.ready()` method (and related `receive()` and `__anext__`
when used as an async iterator) now `raise`s a `ReceiverError` and in
particular a `ReceiverStoppedError` when the receiver has no more
messages to receive.

`Receiver.consume()` doesn't raise any exceptions now (except for
`AssertionError` if there are programming errors).

Receivers raising `EOFError` now raise `ReceiverError` instead.

For channels which receivers stop receiving when the channel is closed,
the `ReceiverStoppedError` is chained with a `__cause__` that is a
`ChannelClosedError` with the channel that was closed.

Since now there is no need to use `ChannelClosedError` without
specifying a channel, now it requires the argument `channel` (before it
was optional).

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
nox: Make output more verbose for darglint

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Move exceptions into its own module

We already have quite a few exceptions, so we better move them to keep
the base classes cleaner.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Add a ReceiverInvalidatedError

This is raised when the receiver was invalidated, for example when it
was converted into a Peekable.

Fixes #52.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Don't raise exceptions in Receiver.ready()

Instead, exceptions are raised in Receiver.consume(). This improves the
API for several reasons:

* There is no need to try: a whole select.ready() loop. Exceptions will
  be raised when something is actually consumed.
* It removes exception handling from tasks monitoring pending receivers.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
Updated to the latest head and added a commit that moves exception raising from `ready()` to `consume()` (previously in #62 only).

ping @sahas-subramanian-frequenz

shsms commented on src/frequenz/channels/_bidirectional.py:
```diff
@@ -23,26 +23,48 @@ class Handle(Sender[V], Receiver[W]):
+                        self._chan,  # pylint: disable=protected-access
+                    )
+                    this_chan_error.__cause__ = err.__cause__
+                    err.__cause__ = this_chan_error
```
don't you have to re-raise the exception here because you're catching it?

leandro-lucarella-frequenz replied:
YUP! Good... *catch* :clown_face:

shsms commented on src/frequenz/channels/_bidirectional.py:
```diff
@@ -67,8 +75,26 @@ async def send(self, msg: V) -> None:
+                    )
+                    this_chan_error.__cause__ = err.__cause__
+                    err.__cause__ = this_chan_error
 
```
missing a raise here too?

leandro-lucarella-frequenz replied:
Yes, but that was fixed when moving the exceptions to `consume()` at least. I'll amend the original location to pretend I never make mistakes anyway :P

shsms commented on src/frequenz/channels/_base_classes.py:
```diff
@@ -133,6 +40,11 @@ async def __anext__(self) -> T:
+        # Break cyclic import
+        from ._exceptions import (  # pylint: disable=import-outside-toplevel,cyclic-import
+            ReceiverStoppedError,
+        )
```
does this have a perf cost,  importing from inside `__anext__`?

It looks like the cyclic importing is necessary just because of type hint dependencies in `_exceptions.py`.  It might be interesting to examine switching to automatically generated typing stubs,  then types become independent:  https://mypy.readthedocs.io/en/stable/stubgen.html

It is hard for me to see the benefit in pulling the exceptions into a separate package,  when we're having to work around ourselves when importing the package.

leandro-lucarella-frequenz replied:
> does this have a perf cost

Good question, my guess is it shouldn't be too much, as import should first check if the module was already imported, so it should be just an extra lookup in the modules dict, which compared to sync primitives should be negligible, but I agree is probably the worse place to break a circular dependency.

I don't think generating stubs is a good idea, it was more of a hack for having at least *some* type info for libs that didn't use native type hints, but if you look at the stubgen example in the link you sent, it doesn't even infer the type of the method arguments.

> Stubgen generates draft stubs. The auto-generated stub files often require some manual updates, and most types will default to `Any`.

> It is hard for me to see the benefit in pulling the exceptions into a separate package, when we're having to work around ourselves when importing the package.

Yeah, for me is just an organizational improvement, is easier to find the code and to avoid noise when looking at unrelated stuff, but I do agree that it might not be the best approach if it needs *workarounding*.

I'll see if there is a better way to avoid the dependency loops and if not I might just move the exceptions back to the base classes module.

leandro-lucarella-frequenz replied:
OK, I found an *elegant* (by Python type hints standards :trollface:) way to break the circular dependency: [`typing.TYPE_CHECKING`](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING).

llucax committed:
test: Properly await for service task

Otherwise pytest will print errors about a dangling task still running.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Make send() raise a SenderError

Instead of returning a `bool`, which can be easily overlooked, it now
raises a new `SenerError`.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

leandro-lucarella-frequenz commented:
[Updated](https://github.com/frequenz-floss/frequenz-channels-python/compare/c390cf27c0497cecde35d21a9de1241927da5c26..c463354124a099a99493155e7ad53f9986b9aa90):

* Remove local import to break circular dependency and use `TYPE_CHECKING` instead.
* Add the missing re-raise statement in `Bidirectional`
* Properly `await` for the service task in the tests to avoid `pytest` errors in `Bidirectional` tests
* Add tests to check exceptions are properly raised in `Bidirectional` (and they are also chained properly)
